### PR TITLE
cuda: Load cuda shared libraries from linked/rpath/LD_LIBRARY_PATH

### PR DESCRIPTION
**Pull request by anustuvicl**
This PR adds the feature that the CUDA component loads the cuda toolkit version that is used by the application.

### CODE REVIEW

File: src/components/cuda/cupti_common.c
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:393]: I would wrap this in a function. E.g., `get_user_cudart_path`

File: src/components/cuda/tests/Makefile
Comment (gcongiu) [inline src/components/cuda/tests/Makefile:26]: What is this?

File: src/components/cuda/cupti_common.c
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:373]: `dl_iterate_phdr_cb` seems a better name.

File: src/components/cuda/cupti_common.c
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:262]: Do you need to cast to void *?

File: src/components/cuda/cupti_common.c
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:260]: Unrelated to this PR. Please never do something like this. Always handle one function at a time.

**Review by gcongiu — CHANGES_REQUESTED** [review]

File: src/components/cuda/tests/Makefile
Comment (anustuvicl) [reply src/components/cuda/tests/Makefile:26]: I made this a comment to have a discussion regarding whether we want this or not.

File: src/components/cuda/cupti_common.c
Comment (anustuvicl) [reply src/components/cuda/cupti_common.c:276]: `linked_cudart_path` is a `const char *`. So freeing it raises a warning `ignores const qualifier`. Casting it to `void *` resolves the warning.

Comment (anustuvicl) [reply src/components/cuda/cupti_common.c:276]: This was also a bug. It shouldn't be freed in `util_load_cuda_sym`. Now it is freed in `cupti_shutdown`

File: src/components/cuda/cupti_common.c
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:403]: This function should also return an integer. Return PAPI_OK

File: src/components/cuda/cupti_common.c
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:392]: return PAPI_OK

File: src/components/cuda/tests/Makefile
Comment (gcongiu) [reply src/components/cuda/tests/Makefile:26]: please remove. if you want to discuss we can do internally.

**Review by gcongiu — CHANGES_REQUESTED** [review]
Also rebase onto master. After you address the comments we can merge this.

File: src/components/cuda/tests/Makefile
Comment (anustuvicl) [reply src/components/cuda/tests/Makefile]: Removed

**Comment by anustuvicl** [comment]
All requested changes applied and rebased.

File: src/components/cuda/cupti_common.c
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:398]: This is a bit awkward as it is written. Can you change it to:
```
if (NULL == linked_cudart_path) {
    return PAPI_EMISC;
}
return PAPI_OK;
```

Comment (anustuvicl) [reply src/components/cuda/cupti_common.c]: Done

**Review by gcongiu — APPROVED** [review]
lets wait for the tests to finish running and then merge

**Pull request merged**
